designs/2022-languages/README.md

```diff
@@ -0,0 +1,703 @@
+
+#### Updating Rule Context
+
+The rule context object, which is passed into rules as `context`, will need to be significantly altered to support other languages. All of the methods that relate specifically to language features will need to move into `SourceCode`, leaving us with a rule context object that follows this interface:
+
+```ts
+interface RuleContext {
+
+    languageOptions: LanguageOptions;
+    settings: object;
+    options: Array<any>;
+    id: string;
+
+    // method replacements
+    cwd: string;
+    filename: string;
+    physicalFilename: string
+    sourceCode: SourceCode;
+
+    report(violation: Violation): void;
+
+    // deprecated methods only available in JS
+    getCwd(): string;
+    getFilename(): string;
+    getPhysicalFilename(): string;
+    getSourceCode(): SourceCode;
+}
+
+interface Violation {
+    messageId: string;
+    data: object;
+    line: number;
+    column: number;
+    endLine?: number;
+    endColumn?: number
+    suggest: Array<Suggestion>;
+    fix(fixer: Fixer): Iterable<FixOp>;
+}
+
+interface Suggestion {
+    desc: string;
+    fix(fixer: Fixer): Iterable<FixOp>;
+}
+```
+
+We should strictly use this interface for all non-JavaScript languages from the start.
+
+**Note:** I have purposely removed the `message` property from the `Violation` interface. I'd like to move to just one way of supplying messages inside of rules and have that be `messageId`. This will also reduce the complexity of processing the violations.
+
+## Documentation
+
+Most existing end-user facing documentation will remain unchanged.
+
+Plugin developer documentation will need to be updated:
+* The documentation on writing new rules needs to be updated to include API changes.
+* A new page on writing languages for ESLint will need to be written. This will need to be quite extensive.
+
+## Drawbacks
+
+1. **Performance.** There will likely be a small performance penalty from this refactoring. Moving to iterators from standard loops seems to incur a slowdown. My hope is that we can mitigate the effects of this performance hit by refactoring code in other ways.
+1. **eslintrc Compatibility.** Unfortunately, there's no easy way to make this functionality available to eslintrc config users. When implemented, eslintrc users will be left without the ability to use other languages.
+
+## Backwards Compatibility Analysis
+
+This proposal is 100% backwards compatible in the short-term, and all breaking changes are optional and can be phased in over time to avoid disrupting the ecosystem.
+
+This proposal does not require any changes from end-users.
+
+## Alternatives
+
+1. We could continue using the `parserForESLint()` approach to support other languages. This was proposed and accepted in [#56](https://github.com/eslint/rfcs/pull/56).
+1. We could choose to not support languages other than JavaScript.
+
+## Open Questions
+
+### Should we eliminate `:exit` from selectors?
```

Comment:
I think this question was more important for earlier versions of this RFC where language implementations were expected to handle selectors. Now that parsing and processing of selectors stays in the ESLint core, this doesn't seem that closely related to the goals of this RFC, so we could evaluate this change separately?

Reply:
Yup, definitely.